modules/configuration/src/main/java/org/apache/ignite/configuration/ConfigurationChanger.java

```diff
@@ -44,68 +47,158 @@
     private final ForkJoinPool pool = new ForkJoinPool(2);
 
     /** Map of configurations' configurators. */
-    private Map<RootKey<?>, Configurator<?>> registry = new HashMap<>();
+    @Deprecated
+    private final Map<RootKey<?>, Configurator<?>> configurators = new HashMap<>();
 
-    /** Storage. */
-    private ConfigurationStorage configurationStorage;
+    /** Map that has all the trees in accordance to their storages. */
+    private final Map<Class<? extends ConfigurationStorage>, StorageRoots> storagesRootsMap = new ConcurrentHashMap<>();
 
-    /** Changer's last known version of storage. */
-    private final AtomicInteger version = new AtomicInteger(0);
+    /**
+     * Immutable data container to store version and all roots associated with the specific storage.
+     */
+    public static class StorageRoots {
+        /** Immutable forest, so to say. */
+        private final Map<RootKey<?>, InnerNode> roots;
+
+        /** Version associated with the currently known storage state. */
+        private final long version;
+
+        /** */
+        private StorageRoots(Map<RootKey<?>, InnerNode> roots, long version) {
+            this.roots = Collections.unmodifiableMap(roots);
+            this.version = version;
+        }
+    }
+
+    /** Storage instances by their classes. Comes in handy when all you have is {@link RootKey}. */
+    private final Map<Class<? extends ConfigurationStorage>, ConfigurationStorage> storageInstances = new HashMap<>();
 
     /** Constructor. */
-    public ConfigurationChanger(ConfigurationStorage configurationStorage) {
-        this.configurationStorage = configurationStorage;
+    public ConfigurationChanger(ConfigurationStorage... configurationStorages) {
+        for (ConfigurationStorage storage : configurationStorages)
+            storageInstances.put(storage.getClass(), storage);
     }
 
     /**
      * Initialize changer.
      */
-    public void init() throws ConfigurationChangeException {
-        final Data data;
+    // ConfigurationChangeException, really?
```

Comment:
Agreed, that was a bad decision? Perhaps, ConfigurationInitException or some ConfigurationException would be more suitable here

Reply:
I mean, why would I get "change" exception from "init" method? Seems weird